### ISSUE TITLE
Nest test log output under test logging tree

### ISSUE THREAD

**Issue by JLLeitschuh** [junit-team/junit5#1127]
## Overview

I may be asking for the moon here due to technological limitations. However, I figure that I should pose this anyways.

I really love the Junit5 tree structured logging output.

However, when you have test logging scattered throughout that code it breaks how it looks.

For example:
```
│  │  ├─ runSSLinkRemovalTest2()
│  │  │       tags: []
│  │  │   uniqueId: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]/[method:runSSLinkRemovalTest2()]
│  │  │     parent: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]
│  │  │     source: MethodSource [className = 'com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest', methodName = 'runSSLinkRemovalTest2', methodParameterTypes = '']
│  │  │   duration: 1 ms
│  │  │     status: ✔ SUCCESSFUL
│  │  ├─ runSSLinkRemovalTest3()
│  │  │       tags: []
│  │  │   uniqueId: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]/[method:runSSLinkRemovalTest3()]
│  │  │     parent: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]
│  │  │     source: MethodSource [className = 'com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest', methodName = 'runSSLinkRemovalTest3', methodParameterTypes = '']
│  │  │   duration: 0 ms
│  │  │     status: ✔ SUCCESSFUL
│  │  ├─ runShortestPathFromSourceTest()
│  │  │       tags: []
│  │  │   uniqueId: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]/[method:runShortestPathFromSourceTest()]
│  │  │     parent: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]
│  │  │     source: MethodSource [className = 'com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest', methodName = 'runShortestPathFromSourceTest', methodParameterTypes = '']
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 1.5    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 3.5    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 6.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 2.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 4.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 5.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
│  │  │   duration: 1 ms
│  │  │     status: ✔ SUCCESSFUL
│  │  ├─ runSSLinkAdditionTest1()
│  │  │       tags: []
│  │  │   uniqueId: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]/[method:runSSLinkAdditionTest1()]
│  │  │     parent: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]
│  │  │     source: MethodSource [className = 'com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest', methodName = 'runSSLinkAdditionTest1', methodParameterTypes = '']
│  │  │   duration: 2 ms
│  │  │     status: ✔ SUCCESSFUL
```

This would look much better if It could be formatted like this instead:

```
│  │  ├─ runShortestPathFromSourceTest()
│  │  │       tags: []
│  │  │   uniqueId: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]/[method:runShortestPathFromSourceTest()]
│  │  │     parent: [engine:junit-jupiter]/[class:com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest]
│  │  │     source: MethodSource [className = 'com.plexxi.fitting.algorithm.shortestpath.FlowAlgorithmsModelTest', methodName = 'runShortestPathFromSourceTest', methodParameterTypes = '']
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 1.5    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 3.5    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 6.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 2.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 4.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 5.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 0.0    
│  │  │   13:02:44.153 [main] DEBUG c.p.f.a.s.FlowAlgorithmsModelTest - 
│  │  │   duration: 1 ms
│  │  │     status: ✔ SUCCESSFUL
│  │  ├─ runSSLinkAdditionTest1()
│  │  │       tags: []
```

I'm assuming that the only way that this would be possible is if you could pass a custom logger implementation into the class under test for this to work correctly but it may be worth it for some people in order to get cleaner log outputs.

## Deliverables

- [ ] Decision on feasibility of hijacking logged output to produce structured output.
- [ ] Implementation of feature.
- [ ] Updated user guide describing how to enable feature

**Comment by sormuras** [comment]
Neat idea.

With messages that are routed through the `org.junit.jupiter.api.TestReporter` we already indent the message and draw the right amount of branches. See https://travis-ci.org/junit-team/junit5/builds/291100399#L1188 and a usage example here http://junit.org/junit5/docs/current/user-guide/#writing-tests-dependency-injection (scroll down a bit).

Either pass the report instance to the class under test or provide a custom logger output handler that routes logging message through the `TestReporter`.

**Comment by sormuras** [comment]
Somehow related issue: #712

**Comment by JLLeitschuh** [comment]
Do you have a suggestion of how to do this with `slf4j`? We aren't using the builtin java logger in our product.
I'm assuming if I configure our `logback.xml` file correctly we could get this to work.
A sample of how to do this in the user guide would be really useful too.

**Comment by sormuras** [comment]
Mh, you can't create an appender/handler at runtime through SLF4J, can you?

LogBack provides a https://logback.qos.ch/manual/appenders.html#SiftingAppender ... which could help here. Or not. Not sure. :)

In a simple project that uses "poor man's logger" `System.out` I achived the goal via:

    @Test
    @AssumeOnline
    void resolve(TestReporter reporter) throws IOException {
      Bach.log.out = (message) -> reporter.publishEntry("message", message);
      [...]

Here's the output:
```
[...]
| +-- BasicsTests [OK]
| | +-- isJavaFile() [OK]
| | +-- tree() [OK]
| | +-- createResolvableFromCoordinates() [OK]
| | +-- resolve(TestReporter) [X] expected: <true> but was: <false>
| | |   2017-10-25T20:26:54.930503300 message = `resolving version from https://oss.sonatype.org/content/repositories/snapshots/org/opentest4j/opentest4j/1.0.0-SNAPSHOT/maven-metadata.xml`
| | |   2017-10-25T20:26:55.377350200 message = `resolved SNAPSHOT as: 20170730.202628-48`
| | |   2017-10-25T20:26:55.377350200 message = `download(uri:https://oss.sonatype.org/content/repositories/snapshots/org/opentest4j/opentest4j/1.0.0-SNAPSHOT/opentest4j-1.0.0-20170730.202628-48.jar, directory:Z:\resolve-7491233623359044718, fileName:opentest4j-1.0.0-SNAPSHOT.jar)`
| | |   2017-10-25T20:26:55.486733800 message = `transferring `https://oss.sonatype.org/content/repositories/snapshots/org/opentest4j/opentest4j/1.0.0-SNAPSHOT/opentest4j-1.0.0-20170730.202628-48.jar`...`
| | |   2017-10-25T20:26:55.800263500 message = `stored `Z:\resolve-7491233623359044718\opentest4j-1.0.0-SNAPSHOT.jar` [2017-07-30T20:26:26Z]`
[...]
```

**Comment by JLLeitschuh** [comment]
There would have to be an API to allow for the configuration of SLF4J prior to any test running for this really to be used practically.
You'd want to have some way to statically get the test reporter before any Junit 5 tests get executed.

**Comment by marcphilipp** [comment]
I think a far more general approach would be to capture stdout/stderr for every test case.

**Comment by marcphilipp** [comment]
Closed in favor of #780.
